feat(core): allow tests to disable the automatic calls of TestBed.res…

oocx opened the pull request:
…etTestingModule in beforeEach

Allow a developer to temporarily disable automatic calls of TestBed.resetTestingModule, so that
a test developer can control when the TestBed should be reset. This enables faster test execution
for some test cases.

## PR Checklist
Does please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->
```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] angular.io application / infrastructure changes
[ ] Other... Please describe:
```

## What is the current behavior?

Angular automatically calls TestBed.resetTestingModule in Jasmine's beforeEach. Because of that behavior, test duration for the tests in our company's project is much longer than necessary.

Issue Number: N/A


## What is the new behavior?

Allow a developer to temporarily disable automatic calls of TestBed.resetTestingModule, so that a test developer can control when the TestBed should be reset. This enables faster test execution for some test cases. For example, a large spec for a complex component that we developed in our company, test execution time was reduced from 45 seconds to 3.5 seconds.

For complex components, we often perform some action in beforeEach and then have many "it" that only test expectations. Today, we need to reexecute the test for every it, although that would not be necessary. The alternative would be to move all expectations to a single "it". However, this would make it much harder to analyze why a test fails. The test would result in a failure after the first failed expectation, and we would not be able to see if the other expectations also failed or if they would have executed successfully.


## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```


## Other information

oocx commented:
What are the next steps for this PR? Do I just need to wait for someone to review this? Am I required to make any changes, so that the travis CI build passes?  I don't understand why the tests fail, so I would require help to do that.

splincode commented:
@oocx what is the status of your pull request? is it still relevant? please fix conflicts if they exist

oocx commented:
@splincode The status is that nobody ever answered my questions and I never got any feedback to this pull request. On gitter I was told that I just had to wait longer until someone got time to look at this PR. 

If the angular team thinks this contribution is valuable and I can get some help with correctly submitting the PR, I'd be happy to continue working on it. In our team we still use a similar solution to speed up our unit tests.

I will have to check if the solution is still valid for the latest Angular release.

kyliau reviewed (approved):
LGTM. Sorry for the long delay. Thank you for the contribution!

JoostK commented on packages/core/testing/src/before_each.ts:
```diff
@@ -19,14 +19,28 @@ declare var global: any;
+ * This is not always necessary and can slow down test execution.
+ * Calling disableTestBedAutoReset disables this behavior.
+ */
+export function disableTestBedAutoReset() {
```
You'll need to actually export these from the public_api file, as these exports would currently not be usable from the outside (as import from `@angular/core/testing`). Doing so warrants a public API golden file update, making it clear that these 2 functions become publicly available.

oocx replied:
How do I do that? I tried "gulp public-api:update", but got an error:

λ gulp public-api:update
[11:49:37] Using gulpfile C:\github\angular\gulpfile.js
[11:49:37] Starting 'build.sh'...
[11:49:38] Finished 'build.sh' after 1.23 s
[11:49:38] Starting 'public-api:update'...
events.js:160
      throw er; // Unhandled 'error' event
      ^

oocx committed:
feat(core): allow tests to disable the automatic calls of TestBed.resetTestingModule in beforeEach

Allow a developer to temporarily disable automatic calls of TestBed.resetTestingModule, so that a
test developer can control when the TestBed should be reset. This enables faster test execution
for some test cases.

jneterer commented:
@oocx This PR would be great. Some serverless SDKs like MongoDB, where you initialize the application in the client, cause serious issues when trying to test services (specifically reinitialization errors) that cannot be avoided in beforeEach. I hope this gets pushed through soon!

AndrewKushnir commented:
Hi @oocx, thank you for the contribution.

I can see how the use-case that you described may require a number of unnecessary recompilations in View Engine TestBed, thus resulting in extra time to complete all tests. However in Ivy TestBed the logic was updated to avoid unnecessary recompilations between test runs, so you should see performance improvement after switching to Ivy.

Given that this behavior was optimized in Ivy, I think we should not extend API surface at this moment.

Thank you.
